### PR TITLE
Address PR comments

### PR DESCRIPTION
**Pull request by dukesteen**
# PR review follow-up plan

> **Status:** Ready for implementation.
> **Baseline:** `master` at `3dbb078` on 2026-07-28. PRs
> [#26](https://github.com/ImmediatePlatform/Immediate.Jobs/pull/26),
> [#33](https://github.com/ImmediatePlatform/Immediate.Jobs/pull/33), and
> [#34](https://github.com/ImmediatePlatform/Immediate.Jobs/pull/34) are merged.
> **Scope:** Resolve all 23 requested review comments against the current code, including provider
> parity and regression coverage.

## Working rules

- Verify each finding against current `master`; do not apply a review suggestion mechanically when
  the code has moved or an existing invariant already addresses it.
- Keep one owner per file during a parallel wave. A task may read any file, but it should edit only
  its listed ownership set unless the coordinator reassigns a file.
- Every behavior change needs a regression test that fails before the change. Provider-specific
  fixes belong in that provider's existing test class; cross-provider invariants belong in the
  relational matrix or functional tests during integration.
- Do not change storage schemas unless a task explicitly requires it. Prefer bounded reads and
  existing indexes before adding durable Redis indexes.
- Subagents should not commit. They should report changed files, tests run, and any finding they
  believe is already fixed or invalid, with evidence.

## Agreed design direction

These decisions remove ambiguity between tasks:

1. Add a bulk incoming-edge read to `IJobGraphStorage`, accepting child job IDs and returning
   `JobContinuationEdge` values. Single-server recovery will call it once per recovered page instead
   of calling `GetJobStatusAsync` per job. Do not put dependency edges on `JobRecord`.
2. `MaxAttempts` is definition metadata, not durable invocation state. Make an unavailable retry
   limit explicit (nullable) at the storage seam; `JobMonitor` continues to fill it from the matching
   registered `JobDefinition`. Never use `0` to mean “unknown.”
3. Empty-batch `FractionSettled` is canonically `1d`: an empty aggregate has no unsettled members.
   Put the calculation in one shared helper and use it in every graph provider.
4. A workflow buffer has an open/sealed lifecycle. Completion atomically seals and snapshots an
   execution buffer before persistence. A batch atomically snapshots jobs and edges before its
   storage call; add/commit/dispose cannot mutate the same lists concurrently.
5. Mixed continuation triggers are evaluated from the complete incoming edge set when the remaining
   dependency count reaches zero. The edge that happened to settle last must not decide the result.
6. The fair-queue demo should use one atomic batch containing the backlog plus a delayed quiet job.
   Add grouped `AddToBatch`/`AddToBatchAt` overloads rather than keeping request-bound orchestration.
7. Redis `QueryJobsAsync` should first make the ID and unfiltered-page paths bounded. Filtered
   queries should scan the ordered all-jobs index in fixed-size windows and stop after collecting
   `Skip + Take` matches. This fixes unbounded materialization without a schema/key migration; add
   secondary ordered indexes only if tests or measurements show the windowed path is insufficient.

## Execution waves

With three subagent slots plus the coordinator, use the following waves. Tasks in the same row have
disjoint ownership and may run in parallel.

| Wave | Parallel tasks | Gate |
| --- | --- | --- |
| 1 | `T01`, `T06`, and the documentation/dashboard/container portions of `T07` | Establish shared contracts while independent work proceeds. |
| 2 | `T02`, `T03`, and `T04` | Start after `T01`; each provider implements the frozen shared contracts. |
| 3 | `T05` and the sample portion of `T07` | Start `T05` after `T02`–`T04`; sample work needs the grouped batch API from `T01`. |
| 4 | `T08` | Integrate all tasks, add only missing matrix coverage, and run the full validation set. |

### T01 — Shared contracts and workflow scheduling safety

**Dependencies:** None.
**Primary ownership:**

- `src/Immediate.Jobs.Shared/IJobStorage.cs`
- `src/Immediate.Jobs.Shared/JobBatchModels.cs`
- `src/Immediate.Jobs.Shared/JobBatches.cs`
- `src/Immediate.Jobs.Shared/JobSchedulers.cs`
- `tests/Immediate.Jobs.FunctionalTests/BatchesAndContinuationsTests.cs`
- focused shared-contract tests added under `tests/Immediate.Jobs.FunctionalTests/`

**Checklist:**

- [x] Add the bulk incoming-edge query to `IJobGraphStorage`, including null/empty/duplicate input
  semantics and XML documentation. An empty ID set should return an empty result.
- [x] Change `JobStatus.MaxAttempts` to represent unknown explicitly and update `JobMonitor` so a
  matching `JobDefinition` always supplies the configured value. Add tests for both a known
  definition and an unknown persisted job.
- [x] Add one shared batch-progress calculation and define the zero-total result as `1d`.
- [x] Make `JobExecutionBuffer.Add` safe for concurrent generated-scheduler calls and replace
  `Snapshot()` with an atomic seal-and-snapshot operation. Additions after sealing must fail with
  a clear `ImmediateJobException`.
- [x] Give `JobBatch` an explicit `Open -> Committing -> Finished/Disposed` lifecycle. Under one
  synchronization gate, validate state, construct immutable job/edge snapshots, and prevent
  concurrent add/commit/dispose mutations. Do not hold a monitor lock across the storage await.
- [x] Validate every standalone continuation parent before any record or edge is created. Reject a
  default handle, null/blank ID, duplicate parent, or invalid batch relationship consistently.
- [x] Add `groupId` overloads for `AddToBatch` and `AddToBatchAt`, using the existing normalization
  rules, so the fair-queue sample can be made atomic.
- [x] Strengthen `AssertBefore` so both events must exist before comparing indexes.
- [x] Cover concurrent batch additions, commit racing an add, concurrent execution-buffer additions,
  post-seal additions, default handles, and grouped batch records.

**Acceptance:** No mutable list can be enumerated while another thread mutates it; storage receives
immutable snapshots; invalid parents fail before persistence; retry limits never surface as `0`
merely because storage cannot know the definition.

**Review comments:**
[r3660721173](https://github.com/ImmediatePlatform/Immediate.Jobs/pull/26#discussion_r3660721173),
[r3660721178](https://github.com/ImmediatePlatform/Immediate.Jobs/pull/26#discussion_r3660721178),
[r3660721185](https://github.com/ImmediatePlatform/Immediate.Jobs/pull/26#discussion_r3660721185),
[r3660721190](https://github.com/ImmediatePlatform/Immediate.Jobs/pull/26#discussion_r3660721190),
[r3660721194](https://github.com/ImmediatePlatform/Immediate.Jobs/pull/26#discussion_r3660721194), and
[r3660721201](https://github.com/ImmediatePlatform/Immediate.Jobs/pull/26#discussion_r3660721201).

### T02 — EF Core graph and fair-queue correctness

**Dependencies:** `T01`.
**Primary ownership:**

- `src/Immediate.Jobs.EntityFrameworkCore/EntityFrameworkCoreJobStorage.cs`
- `tests/Immediate.Jobs.FunctionalTests/Storage/EntityFrameworkCoreJobStorageTests.cs`

**Checklist:**

- [x] Implement the bulk incoming-edge query as one no-tracking query and preserve all edge fields.
- [x] In continuation flush, validate every addition before changing counters or adding entities:
  detached work must have no `BatchId`; tracked work must exactly match the current batch;
  unknown options/triggers must fail. Do not silently rewrite invalid input.
- [x] When the final dependency settles, query/evaluate the child's complete incoming trigger set.
  A mixed Success/Failure graph must resolve identically whether parents were already terminal or
  completed one at a time.
- [x] Use the shared empty-batch progress calculation.
- [x] Return unknown, rather than `0`, for low-level `JobStatus.MaxAttempts`; leave definition
  resolution to `JobMonitor`.
- [x] Move returning-group cursor cleanup into the same execution-strategy/transaction boundary as
  grouped enqueue. Preserve “reset only when the group has no live jobs.”
- [x] Replace `groupStates[job.Id]` with race-tolerant lookup. A vanished candidate should be skipped
  or treated as quiet and retried; it must not throw `KeyNotFoundException`.
- [x] Bound consecutive failed fair-queue claims with a named small limit, reset the counter after a
  successful claim, and exit without spinning when the limit is reached.
- [x] Add tests for invalid continuation batch IDs, both mixed-trigger completion orders, empty-batch
  projection, cursor reset atomicity/semantics, vanished candidate state, and bounded claim loss.

**Acceptance:** EF and in-memory semantics agree for invalid additions and mixed triggers; fair
acquisition terminates under contention and a returning group has no stale cursor debt.

**Review comments:**
[r3660721165](https://github.com/ImmediatePlatform/Immediate.Jobs/pull/26#discussion_r3660721165),
[r3660721178](https://github.com/ImmediatePlatform/Immediate.Jobs/pull/26#discussion_r3660721178),
[r3660721194](https://github.com/ImmediatePlatform/Immediate.Jobs/pull/26#discussion_r3660721194),
[r3660957580](https://github.com/ImmediatePlatform/Immediate.Jobs/pull/33#discussion_r3660957580),
[r3661077751](https://github.com/ImmediatePlatform/Immediate.Jobs/pull/34#discussion_r3661077751),
[r3661077752](https://github.com/ImmediatePlatform/Immediate.Jobs/pull/34#discussion_r3661077752), and
[r3661077756](https://github.com/ImmediatePlatform/Immediate.Jobs/pull/34#discussion_r3661077756).

### T03 — LINQ-to-DB graph and fair-queue correctness

**Dependencies:** `T01`.
**Primary ownership:**

- `src/Immediate.Jobs.LinqToDB/LinqToDBJobStorage.cs`
- `tests/Immediate.Jobs.StorageTests/LinqToDBSqliteStorageTests.cs`

**Checklist:**

- [x] Implement the bulk incoming-edge query in one query.
- [x] Mirror the validated continuation-addition rules from `T02`; do not silently reparent LINQ-to-DB
  additions even though the original comment was anchored on EF.
- [x] Resolve mixed triggers from all incoming edges when the dependency counter reaches zero.
- [x] Use the shared empty-batch progress calculation.
- [x] Return unknown, rather than `0`, for low-level `JobStatus.MaxAttempts`.
- [x] Give grouped enqueue the same transactional returning-group cursor reset semantics as EF.
- [x] Handle missing `groupStates` entries without an indexer exception.
- [x] Bound and reset consecutive failed-claim retries exactly as in EF.
- [x] When no grouped head remains, run the ungrouped fast path, update capacity/acquired results, and
  break instead of reissuing unchanged queries.
- [x] Add SQLite tests for all of the above, including an ungrouped-only queue with fair queues enabled
  and a contention test with a time bound so an infinite loop is observable.

**Acceptance:** LINQ-to-DB terminates every fair-acquisition loop, does not starve ungrouped jobs, and
matches EF/in-memory graph semantics.

**Review comments:**
[r3660721165](https://github.com/ImmediatePlatform/Immediate.Jobs/pull/26#discussion_r3660721165) (provider parity),
[r3660721178](https://github.com/ImmediatePlatform/Immediate.Jobs/pull/26#discussion_r3660721178),
[r3660721194](https://github.com/ImmediatePlatform/Immediate.Jobs/pull/26#discussion_r3660721194),
[r3660957580](https://github.com/ImmediatePlatform/Immediate.Jobs/pull/33#discussion_r3660957580),
[r3661077751](https://github.com/ImmediatePlatform/Immediate.Jobs/pull/34#discussion_r3661077751),
[r3661077752](https://github.com/ImmediatePlatform/Immediate.Jobs/pull/34#discussion_r3661077752),
[r3661077756](https://github.com/ImmediatePlatform/Immediate.Jobs/pull/34#discussion_r3661077756), and
[r3661077757](https://github.com/ImmediatePlatform/Immediate.Jobs/pull/34#discussion_r3661077757).

### T04 — In-memory graph parity and testing assertions

**Dependencies:** `T01`.
**Primary ownership:**

- `src/Immediate.Jobs.Shared/InMemoryJobStorage.cs`
- `src/Immediate.Jobs.Testing/JobTestHarness.cs`
- `tests/Immediate.Jobs.FunctionalTests/Storage/InMemoryJobStorageBatchTests.cs`
- `tests/Immediate.Jobs.FunctionalTests/Packages/TestingPackageTests.cs`

**Checklist:**

- [x] Implement the bulk incoming-edge query under the existing storage lock.
- [x] Use the shared batch-progress helper.
- [x] Return unknown, rather than `0`, for low-level `JobStatus.MaxAttempts`.
- [x] Change incremental dependency settlement to inspect all incoming triggers at zero remaining;
  retain cascade cancellation and settled-edge idempotency.
- [x] Keep the existing strict continuation-addition validation as the provider-parity oracle and add
  explicit tests for detached-with-batch and tracked-with-wrong-batch inputs.
- [x] Make `AssertContinuationReleasedAfterAsync` reject `Cancelled` explicitly, as well as a child
  that remains `AwaitingContinuation`.
- [x] Add mixed-trigger order tests and testing-package tests proving cancellation is not release.

**Acceptance:** The in-memory provider supplies the canonical behavior used by both SQL providers,
and the testing package cannot report a cancelled continuation as released.

**Review comments:**
[r3660721165](https://github.com/ImmediatePlatform/Immediate.Jobs/pull/26#discussion_r3660721165),
[r3660721178](https://github.com/ImmediatePlatform/Immediate.Jobs/pull/26#discussion_r3660721178),
[r3660721194](https://github.com/ImmediatePlatform/Immediate.Jobs/pull/26#discussion_r3660721194),
[r3660721197](https://github.com/ImmediatePlatform/Immediate.Jobs/pull/26#discussion_r3660721197), and
[r3660957580](https://github.com/ImmediatePlatform/Immediate.Jobs/pull/33#discussion_r3660957580).

### T05 — Single-server recovery, telemetry isolation, and disposal

**Dependencies:** `T02`, `T03`, and `T04`.
**Primary ownership:**

- `src/Immediate.Jobs.Shared/SingleServerJobStorage.cs`
- `src/Immediate.Jobs.Shared/JobSchedulerService.cs`
- `tests/Immediate.Jobs.FunctionalTests/Storage/SingleServerJobStorageTests.cs`
- focused scheduler-service tests under `tests/Immediate.Jobs.FunctionalTests/`

**Checklist:**

- [x] During each `QueryJobsAsync` recovery page, bulk-load incoming edges for non-batch jobs, group
  them by child ID, and choose `EnqueueAsync` versus `EnqueueContinuationAsync` without per-job
  durable status calls. Preserve missing batch-header/graph validation.
- [x] Add a counting durable-storage test double and prove recovery calls scale with pages, not jobs.
- [x] Isolate `SetExecutionTelemetryAsync` from the invocation/completion try/catch. Log a warning
  when telemetry persistence fails, but still invoke and complete the job; cancellation caused by
  host shutdown should keep normal cancellation semantics.
- [x] Remove `_initialization.Dispose()` from disposal while retaining the disposal-started guard and
  final release. Exercise the race where callers pass `ThrowIfDisposed` and wait behind disposal.
- [x] Test sync/async/concurrent disposal, waiting initialization callers, telemetry-write failure,
  and successful job completion after that failure.

**Acceptance:** Recovery performs O(pages + batches) durable reads rather than O(jobs); telemetry
failure never consumes a job attempt; no waiter touches a disposed semaphore.

**Review comments:**
[r3660721194](https://github.com/ImmediatePlatform/Immediate.Jobs/pull/26#discussion_r3660721194),
[r3660957620](https://github.com/ImmediatePlatform/Immediate.Jobs/pull/33#discussion_r3660957620), and
[r3661077783](https://github.com/ImmediatePlatform/Immediate.Jobs/pull/34#discussion_r3661077783).

### T06 — Redis bounded queries and action semantics

**Dependencies:** `T01` only for the nullable `MaxAttempts` contract; query/action work may begin
before `T01` and rebase afterward.
**Primary ownership:**

- `src/Immediate.Jobs.Redis/RedisJobStorage.cs`
- `src/Immediate.Jobs.Redis/RedisScripts.cs`
- `tests/Immediate.Jobs.StorageTests/RedisStorageTests.cs`

**Checklist:**

- [x] Keep an exact-ID query O(1).
- [x] For an unfiltered query, range only the requested `AllJobsKey` rank window and deserialize only
  those IDs.
- [x] For state/queue/search filters, scan `AllJobsKey` in fixed-size descending-rank windows,
  deserialize one window at a time, and stop after `Skip + min(Take, 1000)` matches. Preserve
  stable ordering and cancellation between windows.
- [x] Map Lua results consistently: `0` means missing and throws `KeyNotFoundException`; negative
  means the target exists but the requested transition is invalid. Apply this to retry, delete,
  and recurring removal without adding a race-prone existence pre-read.
- [x] Confirm the existing purge script's progress invariant: on missing or state-mismatched jobs it
  still removes the ID from the completed sorted set. Add a regression test with stale completed
  members and a timeout. Add a defensive no-progress guard only if the test exposes a path that
  can return the same batch unchanged.
- [x] Update low-level status construction to return unknown, not `0`, for `MaxAttempts`.
- [x] Add pagination/filter tests with enough jobs to cross multiple windows and not-found tests for
  all three actions.

**Acceptance:** `Take=1` no longer fetches/deserializes the entire job history, filtered pagination is
correct and bounded in memory, missing resources map to dashboard 404 behavior, and purge cannot loop
on stale index entries.

**Review comments:**
[r3660721173](https://github.com/ImmediatePlatform/Immediate.Jobs/pull/26#discussion_r3660721173),
[r3660957586](https://github.com/ImmediatePlatform/Immediate.Jobs/pull/33#discussion_r3660957586),
[r3660957592](https://github.com/ImmediatePlatform/Immediate.Jobs/pull/33#discussion_r3660957592), and
[r3660957602](https://github.com/ImmediatePlatform/Immediate.Jobs/pull/33#discussion_r3660957602).

### T07 — Docs, dashboard, sample, and test-fixture cleanup

**Dependencies:** The sample subtask depends on `T01`; all other subtasks are independent.
**Primary ownership:**

- `docs/fair-queues.md`
- `spec.md`
- `src/Immediate.Jobs.Dashboard/DashboardClient/src/components/JobDetail.vue`
- `src/Immediate.Jobs.Dashboard/DashboardClient/src/components/JobTable.vue`
- `src/Immediate.Jobs.Dashboard/DashboardClient/tests/components.test.ts`
- `samples/Aspire/Api/Endpoints/SampleApiEndpoints.cs`
- related Aspire response/sample tests if added
- `tests/Immediate.Jobs.StorageTests/StorageContainers.cs`

**Checklist:**

- [x] Fix fair-queue pseudo-SQL to use `effectiveGroupId = GroupId ?? Id` for inflight counts, noisy
  detection, and row-number partitioning. Clarify that production providers keep ungrouped jobs
  on the independent/fast path rather than coalescing nulls.
- [x] Mark fair queues implemented in `spec.md` and remove stale “not yet scheduled” and unresolved
  implementation-time wording.
- [x] Use `groupId !== null` consistently in job detail and table rendering. Test both `null` and `""`
  so the `string | null` contract is explicit even though generated schedulers normalize blank IDs.
- [x] Replace the fair-queue sample endpoint's 100 request-bound writes and cancellable delay with a
  single atomic batch: immediate backlog members plus one quiet member delayed five seconds, all
  carrying their intended group IDs. Return accepted only after commit succeeds.
- [x] Dispose PostgreSQL and SQL Server test containers concurrently with `Task.WhenAll` (or an
  equivalent aggregate/finally pattern) so both disposal attempts happen when either one throws.
- [x] Run dashboard lint, typecheck, component tests, and build after the Vue changes.

**Acceptance:** Documentation matches the shipped algorithm/status, empty group IDs render according
to the declared contract, an aborted sample request cannot leave a partial demo run, and fixture
cleanup attempts both containers.

**Review comments:**
[r3660957561](https://github.com/ImmediatePlatform/Immediate.Jobs/pull/33#discussion_r3660957561),
[r3660957635](https://github.com/ImmediatePlatform/Immediate.Jobs/pull/33#discussion_r3660957635),
[r3661077735](https://github.com/ImmediatePlatform/Immediate.Jobs/pull/34#discussion_r3661077735),
[r3661077743](https://github.com/ImmediatePlatform/Immediate.Jobs/pull/34#discussion_r3661077743), and
[r3661077746](https://github.com/ImmediatePlatform/Immediate.Jobs/pull/34#discussion_r3661077746).

### T08 — Integration, provider matrix, and review closeout

**Dependencies:** `T01`–`T07`.
**Primary ownership:** coordinator-owned integration changes only; avoid reopening task-owned files
unless a failing cross-provider test requires it.

**Checklist:**

- [x] Add cross-provider matrix coverage for the invariants that should match: bulk incoming edges,
  mixed triggers, invalid dynamic-addition batch IDs, empty-batch progress, and missing-resource
  action behavior where the storage contract requires it.
- [x] Audit all 23 links in the coverage table below. For any comment closed as already safe, record
  the exact code/test evidence in the eventual PR response.
- [x] Run targeted project tests first, then the full repository validation listed below.
- [x] Inspect `git diff --check`, generated/dashboard assets, public API implications of nullable
  `MaxAttempts`, and packaging/AOT build results.
- [x] Prepare one concise response per GitHub thread with the implementing test or the evidence for
  no production change. Do not resolve/post comments unless explicitly authorized.

**Acceptance:** Every requested thread has code-and-test evidence or a documented no-change reason;
the full build, tests, dashboard checks, NativeAOT publish, and pack succeed.

## Review-comment coverage

| Comment | Finding | Tasks |
| --- | --- | --- |
| [r3660721165](https://github.com/ImmediatePlatform/Immediate.Jobs/pull/26#discussion_r3660721165) | EF silently reparents invalid continuation additions | `T02`, parity in `T03`/`T04` |
| [r3660721173](https://github.com/ImmediatePlatform/Immediate.Jobs/pull/26#discussion_r3660721173) | `MaxAttempts` hard-coded to zero | `T01`–`T04`, `T06` |
| [r3660721178](https://github.com/ImmediatePlatform/Immediate.Jobs/pull/26#discussion_r3660721178) | Empty-batch progress differs | `T01`–`T04` |
| [r3660721185](https://github.com/ImmediatePlatform/Immediate.Jobs/pull/26#discussion_r3660721185) | Unsynchronized workflow buffers | `T01` |
| [r3660721190](https://github.com/ImmediatePlatform/Immediate.Jobs/pull/26#discussion_r3660721190) | Default continuation handles accepted | `T01` |
| [r3660721194](https://github.com/ImmediatePlatform/Immediate.Jobs/pull/26#discussion_r3660721194) | Single-server recovery N+1 | `T01`–`T05` |
| [r3660721197](https://github.com/ImmediatePlatform/Immediate.Jobs/pull/26#discussion_r3660721197) | Cancelled child treated as released | `T04` |
| [r3660721201](https://github.com/ImmediatePlatform/Immediate.Jobs/pull/26#discussion_r3660721201) | Vacuous `AssertBefore` | `T01` |
| [r3660957561](https://github.com/ImmediatePlatform/Immediate.Jobs/pull/33#discussion_r3660957561) | Docs group null IDs together | `T07` |
| [r3660957580](https://github.com/ImmediatePlatform/Immediate.Jobs/pull/33#discussion_r3660957580) | Mixed triggers depend on last edge | `T02`–`T04` |
| [r3660957586](https://github.com/ImmediatePlatform/Immediate.Jobs/pull/33#discussion_r3660957586) | Redis query materializes all jobs | `T06` |
| [r3660957592](https://github.com/ImmediatePlatform/Immediate.Jobs/pull/33#discussion_r3660957592) | Redis missing-resource semantics | `T06` |
| [r3660957602](https://github.com/ImmediatePlatform/Immediate.Jobs/pull/33#discussion_r3660957602) | Possible Redis purge loop | `T06` (existing Lua invariant + regression) |
| [r3660957620](https://github.com/ImmediatePlatform/Immediate.Jobs/pull/33#discussion_r3660957620) | Telemetry failure aborts execution | `T05` |
| [r3660957635](https://github.com/ImmediatePlatform/Immediate.Jobs/pull/33#discussion_r3660957635) | First container disposal can leak second | `T07` |
| [r3661077735](https://github.com/ImmediatePlatform/Immediate.Jobs/pull/34#discussion_r3661077735) | Fair demo leaves partial runs | `T01`, `T07` |
| [r3661077743](https://github.com/ImmediatePlatform/Immediate.Jobs/pull/34#discussion_r3661077743) | Fair-queue roadmap is stale | `T07` |
| [r3661077746](https://github.com/ImmediatePlatform/Immediate.Jobs/pull/34#discussion_r3661077746) | Dashboard group presence differs from contract | `T07` |
| [r3661077751](https://github.com/ImmediatePlatform/Immediate.Jobs/pull/34#discussion_r3661077751) | Cursor reset is EF-only/non-transactional | `T02`, `T03` |
| [r3661077752](https://github.com/ImmediatePlatform/Immediate.Jobs/pull/34#discussion_r3661077752) | Candidate state can disappear | `T02`, `T03` |
| [r3661077756](https://github.com/ImmediatePlatform/Immediate.Jobs/pull/34#discussion_r3661077756) | Failed claims retry forever | `T02`, `T03` |
| [r3661077757](https://github.com/ImmediatePlatform/Immediate.Jobs/pull/34#discussion_r3661077757) | LINQ-to-DB spins when grouped heads vanish | `T03` |
| [r3661077783](https://github.com/ImmediatePlatform/Immediate.Jobs/pull/34#discussion_r3661077783) | Disposed semaphore strands/faults waiters | `T05` |

## Validation commands

Run targeted tests within each task, then run the complete set from the repository root:

```bash
dotnet restore
dotnet build -c Release --no-restore
dotnet test -c Release --no-build
dotnet publish samples/NativeAot/NativeAot.csproj -c Release -r linux-x64 -f net8.0 --self-contained --no-restore
dotnet pack -c Release --no-build
```

Run dashboard checks from `src/Immediate.Jobs.Dashboard/DashboardClient`:

```bash
npm ci
npm run check
npm run build
npm run size
```

The storage test project requires Docker for Redis, PostgreSQL, and SQL Server Testcontainers. Also
run `git diff --check` before handoff.

## Definition of done

- [x] All `T01`–`T08` acceptance criteria are satisfied.
- [x] Each of the 23 review comments maps to a passing regression test or a written, code-backed
  no-change explanation.
- [x] EF, LINQ-to-DB, and in-memory behavior matches for graph invariants.
- [x] Fair acquisition has a finite exit path for empty candidate sets and repeated claim loss.
- [x] Redis job queries use bounded windows and action methods distinguish missing from invalid state.
- [x] Full CI-equivalent validation passes without touching unrelated workspace files.

## Summary by CodeRabbit

* **New Features**
  * Added group-aware batch scheduling overloads (including delayed/scheduled variants) and support for fetching incoming continuation edges.
  * Updated the fair-queue demo endpoint to showcase atomic grouped backlog rotation behavior.
* **Bug Fixes**
  * Dashboard now correctly renders Group when the group id is not null (including falsy-but-valid values like `0` and `''`).
  * Empty batch progress reports fully settled; job status now reports `MaxAttempts` as unknown (`null`) when not determinable.
* **Improvements**
  * Strengthened fairness/continuation validation and retry/delete semantics; added query filtering by job name.
* **Documentation**
  * Refined batching, fair-queue, and related API documentation/spec wording to match current contracts.

### CODE REVIEW

File: src/Immediate.Jobs.Dashboard/Assets/app.js
Comment (viceroypenguin) [inline src/Immediate.Jobs.Dashboard/Assets/app.js:1]: Should this file be checked in at all? If it is a file generated from the build process, then it should be marked in `.gitignore` and not checked in.

File: src/Immediate.Jobs.Shared/JobBatchModels.cs [hunk 1]
```diff
@@ -142,11 +142,32 @@ public sealed record JobContinuationAddition
 /// <summary>Per-attempt buffer used by generated schedulers during job execution.</summary>
 public sealed class JobExecutionBuffer
 {
+	private readonly Lock _gate = new();
 	private readonly List<JobContinuationAddition> _additions = [];
+	private bool _sealed;
```
[the remark this answers is not in the document]
Comment (viceroypenguin) [reply src/Immediate.Jobs.Shared/JobBatchModels.cs:174]: `System.Threading.Lock` is polyfilled, so this compiles cleanly

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -46,12 +46,6 @@ jobs:
       run: npm run check
       working-directory: src/Immediate.Jobs.Dashboard/DashboardClient
 
-    - name: Build dashboard
```
Comment (viceroypenguin) [inline .github/workflows/build.yml:49]: note to self: this is moved to `dotnet build` implicitly

**Review by viceroypenguin — APPROVED** [review]
